### DISCUSSION TITLE
Understanding nginx-proxy routing behavior

### DISCUSSION THREAD

**Discussion by kzhmr1** [nginx-proxy/nginx-proxy#1964] [Q&A]
Can you pls. explain whether what we are observing is the expected behavior ? This is the scenario, we have two applications, we call them blue/green. The applications are deployed using Ansible yaml file.  Here are the steps that we did:
1. Deploy application green.
2. Using a test client that sends request every second, we observed nginx is successfully routing the request to application green.
3. Deploy application blue.
4. It takes nginx almost 224 seconds to start routing request to application blue, from then on, it alternates sending to blue and green.
5. Stop both applications and do the same step 1 & 3.
6. This time, nginx route to blue application only after 25 seconds.

Question: Why does it take nginx so long the first time to route traffic to blue ?  Please help us understand this behavior so that we can explain this to our application team. Thanks.

**Comment by wader** [comment]
Are green and blue two different containers of the same application with the same `VIRTUAL_HOST` value?
What HTTP client is used for testing and how? could keepalive make results look strange?

**Comment by kzhmr1** [comment]
Basically they are the same application, it's just a helloworld
application. With blue application, I just added more jar files and import
statement to make it load slower. But if I just blue application without
using nginx as a standalone container, the application loads within 29
seconds.

I am just using a basic curl command like this:
while true; do curl -s [url]|grep Hello; sleep 1; done

**Comment by kzhmr1** [comment]
Also, an reply to your question, they are 2 separate containers using the
same VIRTUAL HOST

**Comment by pitkley** [comment]
If you start multiple containers with the same `VIRTUAL_HOST`, nginx-proxy adds multiple servers to the relevant upstream:

``` console
# docker run -d --name np_test -v /var/run/docker.sock:/tmp/docker.sock:ro jwilder/nginx-proxy
# docker run -d --name blue -e VIRTUAL_HOST=test.localhost nginx
# docker run -d --name green -e VIRTUAL_HOST=test.localhost nginx

# docker exec np_test cat /etc/nginx/conf.d/default.conf
...
upstream test.localhost {
                        # green
                        server 172.17.0.5:80;
                        # blue
                        server 172.17.0.4:80;
}
server {
        server_name test.localhost;
        listen 80 ;
        access_log /var/log/nginx/access.log vhost;
        location / {
                proxy_pass http://test.localhost;
        }
}
...
```

Please verify that you see a similar configuration. If you do, your potential "issues" are with nginx and not nginx-proxy (see the [nginx documentation on load balancing](http://nginx.org/en/docs/http/load_balancing.html) for how nginx behaves in a case like this).

If the configuration generated by your nginx-proxy differs vastly and seems incorrect, it might be an issue with nginx-proxy after all.
